### ISSUE TITLE
Transformer training doesn't converge on byte-LM (V=256, d=64-128) — 519s on CPU produces 0% top-1, PPL=V (flat softmax)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1232]
## Summary

After upgrading to AiDotNet **0.172.0** + Tensors **0.69.4**, training a `TransformerArchitecture<float>` on a standard byte-LM task via the `AiModelBuilder` facade for 3 epochs produces an output distribution indistinguishable from uniform random: **top-1 = 0.00%, PPL = 256 (= V)**, after **519 seconds of CPU training** (`GpuUsageLevel.AlwaysCpu` to work around AiDotNet.Tensors#285).

This reproduces across two independent test sites: my new `Phase_L4_AssociativeRetrieval_HeadToHead` H2H test AND the existing `Phase_A4_TransformerBaseline_1MB` baseline test. Both are using the same facade pattern. Both produce the same flat-softmax result.

This is distinct from the prior #1227 (GradientTape Clone leak) and #1228 (single-threaded train) bugs — those are runtime crash + perf bugs, this is a CONVERGENCE bug. The model trains to flat-softmax instead of a learned distribution.

## Reproduction

Existing test in HarmonicEngine repo (no changes needed since v0.166.0):

`Phase_A4_TransformerBaseline_1MB.TransformerBaseline_Cycle30Repro_1MBShakespeare`

Config: V=256, ctx=32, d_model=64, ff=256, heads=4, layers=2, Adam LR=0.0003, epochs=3, stride=16, 62,738 training samples, 500 eval positions.

Output (`logs/Phase_A4_TransformerBaseline_1MB.log`):
```
Phase A4 — Transformer baseline on 1MB Shakespeare — config 'Cycle30Repro'
  V=256 ctx=32 d_model=64 ff=256 heads=4 layers=2
  Adam LR=0.0003 epochs=3 stride=16
  AiDotNet 0.166.0 (post-#1187/#1208 fixes)
...
  Training samples: 62,738
  Training wall: 519.3s

  ===== Cycle30Repro RESULT =====
    val top-1:   0.00%  (F_HRRBind ref: 30.20%)
    val top-5:   0.00%
    val PPL:    256.00  (F_HRRBind ref: 14.63, KN-5gram ceiling: 5.44)
    nll/tok:    5.5452
```

`nll/tok = 5.5452 = ln(256)` exactly — meaning the predicted distribution is **uniform across all 256 byte classes**. The model trained for 519 seconds on CPU and produced no learned signal.

Same symptom in `Phase_L4_AssociativeRetrieval_HeadToHead` at d=64 (my new test, 2 seeds, ~960 sec total CPU train wall):
```
ClassicalTransformer seed=0: top1=0.00% PPL=256.38 params=132,608 train=450.0s
ClassicalTransformer seed=1: top1=0.00% PPL=256.29 params=132,608 train=513.4s
```

Note PPL is 256.38 ± 0.x rather than exactly 256 — the model's logits are **not exactly uniform** but they're so close to uniform that argmax retrieval is at chance. This suggests the gradient updates ARE being applied but the loss surface is being navigated incorrectly (e.g., gradient direction inverted, or the loss function is being computed against the wrong target shape).

## Reference numbers from the same eval protocol

Same corpus, same eval positions, same train/val split, different model architectures via `ILMPredictor`:

| Architecture | Top-1 | PPL | Params | Train wall |
|---|---|---|---|---|
| AiDotNet Transformer (this bug) | **0.00%** | **256.0** | 132,608 | 519s |
| Hand-rolled MinimalTransformerPredictor | ~26% | ~16 | similar | ~30s |
| F_HRRBind cycle-44b champion | 30.20% | 14.63 | 65,792 | <5s |
| KneserNey 5-gram (ceiling) | 51.40% | 5.44 | n/a | n/a |
| AssociativeRetrievalSelfAttn | **36-38%** | **60-77** | **0** | **<0.1s** |

The hand-rolled `MinimalTransformerPredictor` (same architecture: 2-layer multi-head softmax attention + FFN + LayerNorm + tied output head, Adam optimizer) trains to ~26% top-1 in ~30s. The AiDotNet facade with the same hyperparameters trains to 0% top-1 in 519s. Same task, same data, same training budget, different result.

## Hypotheses for root cause

1. **Loss function output-shape mismatch.** `CategoricalCrossEntropyLoss<float>` expects `labels [N, V]` one-hot AND model output `[N, V]` logits. If the model is outputting `[N, ctxLen, V]` (sequence-of-classifications) and the loss is comparing slot-by-slot to a single one-hot `[N, V]`, you'd get a near-flat gradient.

2. **Embedding-tying mis-applied.** If the output projection is tied to the input embedding but applied with the wrong sign/transpose, the gradient might point AWAY from the target instead of toward it.

3. **Gradient clipping / Adam config bug.** Default Adam settings may be silently clipping gradients to ~0 or have an inverted update direction.

4. **Data loader format mismatch.** `DataLoaders.FromTensors<float>(features, labels)` may be producing unexpected shapes that the Transformer interprets as a sequence-classification rather than next-token-prediction.

## Severity

**P0** — blocks the standard "AiDotNet for classical baselines" workflow. Per memory `feedback_use_aidotnet_for_baselines.md` (HarmonicEngine project): use AiDotNet for classical NN baselines in head-to-head tests. With this bug, that's not possible — Transformer training silently produces useless models.

The hand-rolled `MinimalTransformerPredictor` is currently the only working classical-Transformer baseline in HarmonicEngine. AiDotNet's Transformer.BuildAsync does not converge on byte-LM tasks at any d ∈ {64, 128} I've tested.

## Workaround currently in use

`MinimalTransformerPredictor` (hand-rolled) is the substitute classical baseline for HarmonicEngine PR #105's L4 head-to-head until this issue lands. Same architecture, same Adam optimizer, same task — different result. The PR description points reviewers at this issue.

## Asks

1. Verify the bug reproduces on a different machine (not just my AMD CPU build).
2. Trace one training step's loss + grad through `Transformer.Train` — confirm gradients are non-zero AND point in a sensible direction relative to the target one-hot.
3. Check whether `CategoricalCrossEntropyLoss<float>` agrees with `TransformerArchitecture`'s output shape.

**Comment by ooples** [comment]
Fix is up in #1231. Root cause was that token-input transformer architectures (`vocabularySize > 0`) were getting a `GlobalPoolingLayer<T>(Average)` inserted before the classification head, which mean-pools over the sequence axis. For real-data byte-LM training every prefix collapses to roughly the same averaged hidden state and softmax converges to `~uniform / V` — exactly the `nll/tok = ln(V) = 5.5452` and `PPL = V` symptom you reported.

PR #1231 adds `SequencePoolingMode { LastToken, MeanPool, ClsToken, None }` and defaults to `LastToken` for `vocabularySize > 0` (canonical autoregressive LM contract — matches GPT / Llama / Mistral). The dispatch goes through a new `SequenceTokenSliceLayer<T>` that uses tape-tracked `TensorSliceAxis` so gradients correctly flow only to the selected position.

Three integration tests guard the fix:
1. Architecture-default contract (vocab > 0 → LastToken, vocab == 0 → MeanPool, explicit override honoured)
2. `SequenceTokenSliceLayer` forward correctness for `Position.Last` / `Position.First` plus rank validation
3. End-to-end convergence smoke test on a deterministic byte-LM task — 5/5 stability runs converged to NLL < 0.01 / top-1 8/8

Closes once #1231 merges.

**Comment by ooples** [comment]
## Verified FIXED at AiDotNet 0.184.0 (2026-05-09)

Re-ran the original #1187 reproducer (`Phase_A4_AiDotNet_1187_Reproducer.cs`) plus a stronger 5000-iter convergence check at AiDotNet 0.184.0. **Plateau bug genuinely landed.**

### Evidence

500-iter run on identity task (V=8, seqLen=4, d=16, 1 layer):
- Loss varies (no longer bit-identical): 4.52 → 1.95
- 3/8 identity acc = 38% (vs 12.5% chance) — partial convergence

5000-iter run, same config:
- Loss 1.09 → 0.23 → 0.0034 → **0.0000** (monotonic)
- Final-200 avg loss: 0.0114
- Identity eval: **8/8 = 100% accuracy**
- Eval CE: 1.274 < random ln(8)=2.079

### Conclusion

`Transformer<float>.Train()` at AiDotNet 0.184.0 fully converges on the trivial identity task that exhibited the original plateau. The fix chain #1187 → #1191 → #1208 → #1221 → #1232 collectively resolved the bug. Closing this comment thread on the AiDotNet side.

Downstream HarmonicEngine impact: Phase A4 baseline (d=256+ trained Transformer for paper-grade vs comparison) is now unblocked. Will validate at d=64/V=256/Tolstoy 1MB scale next.

Reproducer files in HarmonicEngine repo:
- `tests/Learning/Phase_A4_AiDotNet_1187_Reproducer.cs` (original)
- `tests/Learning/Phase_PA_Transformer_ConvergenceCheck.cs` (longer training, full converge demo)
